src/Controller/Component/PaginatorComponent.php

```diff
@@ -299,13 +299,23 @@ public function getDefaults($alias, $defaults)
         if (isset($defaults[$alias])) {
             $defaults = $defaults[$alias];
         }
-        if (isset($defaults['limit']) &&
-            (empty($defaults['maxLimit']) || $defaults['limit'] > $defaults['maxLimit'])
-        ) {
-            $defaults['maxLimit'] = $defaults['limit'];
+
+        $config = $this->config();
+        $maxLimit = isset($defaults['maxLimit']) ? $defaults['maxLimit'] : $config['maxLimit'];
+        $limit = isset($defaults['limit']) ? $defaults['limit'] : $config['limit'];
+
+        if ($limit > $maxLimit) {
+            $limit = $maxLimit;
         }
+        if (empty($defaults['limit'])) {
+            $defaults['maxLimit'] = $maxLimit;
+            $defaults['limit'] = $limit;
+        }
+
+        $defaults['maxLimit'] = $maxLimit;
+        $defaults['limit'] = $limit;
 
-        return $defaults + $this->config();
+        return $defaults + $config;
```

Comment:
The code should **never** have something like `$defaults + $config`.
As $config by definition should be higher than $defaults in priority of merging. And is exactly the opposite here.
I think that may play into it here. Either way we need to fix the variables here to avoid future confusion and issues.